library/core/src/fmt/mod.rs

```diff
@@ -277,55 +279,111 @@ impl<'a> Formatter<'a> {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[derive(Copy, Clone)]
 pub struct Arguments<'a> {
-    // Format string pieces to print.
-    pieces: &'a [&'static str],
+    /// The number of string pieces and place holders combined.
+    ///
+    /// For example:
+    ///  - 1 for `format_args!("abc")`
+    ///  - 2 for `format_args!("abc{}")`
+    ///  - 3 for `format_args!("abc{}xyz")`
+    ///  - 4 for `format_args!("abc{}xyz{}")`
+    ///  - 5 for `format_args!("abc{}xyz{}123")`
+    ///
+    /// The first part is always a string piece, but it may be an empty string.
+    /// E.g. format_args!("{}") has two parts, one empty string piece and one placeholder.
+    ///
+    /// The number of placeholders is `num_parts / 2`.
+    /// The number of string pieces is `(num_parts + 1) / 2`.
+    num_parts: NonZeroUsize,
+
+    /// The string pieces and the placeholders.
+    ///
+    /// If `num_parts` is one, this stores the &'static str directly.
+    /// Otherwise, it stores pointers to both the slice of string pieces and the slice of placeholders.
+    ///
+    /// The length of those slices are determined by the `num_parts` field above.
+    parts: Parts,
 
-    // Placeholder specs, or `None` if all specs are default (as in "{}{}").
-    fmt: Option<&'a [rt::Placeholder]>,
+    /// Pointer to the start of the array of arguments.
+    args: *const rt::Argument<'a>,
+}
 
-    // Dynamic arguments for interpolation, to be interleaved with string
-    // pieces. (Every argument is preceded by a string piece.)
-    args: &'a [rt::Argument<'a>],
+#[derive(Copy, Clone)]
+union Parts {
+    /// Used if `num_parts == 1`.
+    string: &'static str,
+    /// Used if `num_parts > 1`.
+    ///
+    /// The placaeholders pointer can be null for default placeholders:
+    /// a placeholder for each argument once, in order, with default formatting options.
+    strings_and_placeholders: (*const &'static str, *const rt::Placeholder),
 }
```

Comment:
Why not use an enum? A variant for the `num_parts == 1` case, and the other for the general case.
That's how the type is used AFAIU.